File: android/app/src/main/java/com/tappyai/app/reviews/ui/ReviewsFeedViewModel.kt
package com.tappyai.app.reviews.ui

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.navigation.toRoute
import com.tappyai.app.reviews.data.Review
import com.tappyai.app.reviews.data.ReviewContentType
import com.tappyai.app.reviews.data.ReviewErrorMessages
import com.tappyai.app.reviews.data.ReviewFeedType
import com.tappyai.app.reviews.data.ReviewsRepository
import com.tappyai.features.auth.data.AuthRepository
import com.tappyai.core.logging.LoggerProvider
import com.tappyai.core.network.NetworkResult
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.sync.withPermit
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlin.math.round
import kotlin.math.roundToInt
import javax.inject.Inject

/**
 * State for the vertical reviews feed. [reviews] is the accumulated, paginated list the pager
 * renders; the boolean flags drive the loading / empty / error overlays the screen shows around
 * the pager.
 */
data class ReviewsFeedUiState(
    val reviews: List<Review> = emptyList(),
    val isInitialLoading: Boolean = false,
    val isLoadingMore: Boolean = false,
    val error: String? = null,
    val endReached: Boolean = false,
    /** Active feed tab — For You (default) / Following / Latest, mirroring the web. */
    val feedType: ReviewFeedType = ReviewFeedType.ForYou,
    // The signed-in user's id (JWT `sub`) so the feed can show the own-post overflow (delete/hide)
    // on the user's OWN posts only, mirroring the web feed. Null when signed out.
    val currentUserId: String? = null,
)

/**
 * Where the pager's rows come from. [Explore] is the discovery feed (the tabs' sort/following);
 * [Profile] is one profile's clips — the signed-in user's own posts when [Profile.userId] is null,
 * otherwise that author's — the same calls the profile grids make, so the pager pages exactly the
 * grid, in the grid's order, and never a discovery row.
 */
sealed interface ReviewsFeedSource {
    data object Explore : ReviewsFeedSource
    data class Profile(val userId: String?) : ReviewsFeedSource
    /**
     * The signed-in user's saved reviews — the self profile's "Đã lưu" grid, in its order (newest
     * save first). Self-only by the route's construction (`GET /api/reviews/saved` keys on the
     * bearer); one unpaged call, each reduced row hydrated through `GET /api/reviews/{id}`.
     */
    data object Saved : ReviewsFeedSource
}

@HiltViewModel
class ReviewsFeedViewModel @Inject constructor(
    private val repository: ReviewsRepository,
    private val authRepository: AuthRepository,
    private val logger: LoggerProvider,
    private val reviewErrorMessages: ReviewErrorMessages,
    savedStateHandle: SavedStateHandle,
) : ViewModel() {

    // The Feed destination has no arguments; ProfileClips carries the profile. Read once.
    val source: ReviewsFeedSource = sourceFrom(savedStateHandle)

    private val _uiState = MutableStateFlow(ReviewsFeedUiState(currentUserId = authRepository.currentUserId()))
    val uiState: StateFlow<ReviewsFeedUiState> = _uiState.asStateFlow()

    private var page = 0
    private var loadJob: Job? = null

    init {
        loadFirstPage()
    }

    /** (Re)load from page 0, replacing the list. Used for initial load, retry, and refresh. */
    fun refresh() = loadFirstPage()

    /** Switches the active feed tab and reloads from page 0. Mirrors the web, which aborts the
     *  in-flight request and refetches on tab change; clearing [ReviewsFeedUiState.reviews] makes
     *  the loading overlay show for the newly-selected tab rather than the old tab's posts. */
    fun onFeedTypeChange(type: ReviewFeedType) {
        if (_uiState.value.feedType == type) return
        _uiState.update { it.copy(feedType = type, reviews = emptyList()) }
        loadFirstPage()
    }

    private fun loadFirstPage() {
        loadJob?.cancel()
        page = 0
        val type = _uiState.value.feedType
        _uiState.update { it.copy(isInitialLoading = true, error = null, endReached = false) }
        loadJob = viewModelScope.launch {
            when (val result = loadPage(0, type)) {
                is NetworkResult.Success -> _uiState.update {
                    it.copy(
                        reviews = result.data,
                        isInitialLoading = false,
                        error = null,
                        // A profile's own-posts list and the saved list are one unpaged call each;
                        // the others page by size.
                        endReached = source.isMine || result.data.size < PAGE_SIZE,
                    )
                }
                is NetworkResult.Error -> {
                    logger.e(TAG, "Feed load failed: ${result.error}")
                    _uiState.update {
                        it.copy(isInitialLoading = false, error = reviewErrorMessages.toUserMessage(result.error))
                    }
                }
            }
        }
    }

    /**
     * Called by the pager as the user approaches the end of the loaded list. Loads and appends the
     * next page. Guarded so overlapping settles, an in-flight load, or a reached end are no-ops.
     */
    fun onPageSettled(index: Int) {
        val state = _uiState.value
        if (state.isInitialLoading || state.isLoadingMore || state.endReached) return
        if (index < state.reviews.size - PREFETCH_DISTANCE) return
        loadNextPage()
    }

    private fun loadNextPage() {
        val type = _uiState.value.feedType
        _uiState.update { it.copy(isLoadingMore = true) }
        viewModelScope.launch {
            when (val result = loadPage(page + 1, type)) {
                is NetworkResult.Success -> {
                    page += 1
                    // De-dupe by id: the feed has no cursor, so a row inserted between page
                    // requests can shift the window and repeat an item — a duplicate key would
                    // crash the pager's keyed items.
                    val existingIds = _uiState.value.reviews.mapTo(HashSet()) { it.id }
                    val fresh = result.data.filter { it.id !in existingIds }
                    _uiState.update {
                        it.copy(
                            reviews = it.reviews + fresh,
                            isLoadingMore = false,
                            endReached = result.data.size < PAGE_SIZE,
                        )
                    }
                }
                is NetworkResult.Error -> {
                    // Keep what's already shown; just stop the spinner. Next settle retries.
                    logger.e(TAG, "Feed page ${page + 1} failed: ${result.error}")
                    _uiState.update { it.copy(isLoadingMore = false) }
                }
            }
        }
    }

    /** Optimistically flips like state + count, reverting if the backend call fails. */
    fun toggleLike(review: Review) {
        val target = !review.likedByMe
        updateReview(review.id) {
            it.copy(likedByMe = target, likeCount = (it.likeCount + if (target) 1 else -1).coerceAtLeast(0))
        }
        viewModelScope.launch {
            val result = repository.toggleLike(review.id)
            if (result is NetworkResult.Error) {
                updateReview(review.id) {
                    it.copy(likedByMe = review.likedByMe, likeCount = review.likeCount)
                }
            } else if (result is NetworkResult.Success) {
                // Reconcile with the server's authoritative state.
                updateReview(review.id) { it.copy(likedByMe = result.data) }
            }
        }
    }

    /** Optimistically flips save state + count, reverting if the backend call fails. */
    fun toggleSave(review: Review) {
        val target = !review.savedByMe
        updateReview(review.id) {
            it.copy(savedByMe = target, saveCount = ((it.saveCount ?: 0) + if (target) 1 else -1).coerceAtLeast(0))
        }
        viewModelScope.launch {
            val result = repository.toggleSave(review.id)
            if (result is NetworkResult.Error) {
                updateReview(review.id) {
                    it.copy(savedByMe = review.savedByMe, saveCount = review.saveCount)
                }
            } else if (result is NetworkResult.Success) {
                updateReview(review.id) { it.copy(savedByMe = result.data) }
            }
        }
    }

    /**
     * Deletes the caller's own post (web parity: own-post overflow → Delete). Optimistically removes
     * it from the feed; a failure re-adds the prior list. Server enforces ownership independently.
     */
    fun deleteReview(review: Review) {
        val previous = _uiState.value.reviews
        _uiState.update { it.copy(reviews = it.reviews.filterNot { r -> r.id == review.id }) }
        viewModelScope.launch {
            val result = repository.deleteReview(review.id)
            if (result is NetworkResult.Error) {
                logger.e(TAG, "Delete review failed: ${result.error}")
                _uiState.update { it.copy(reviews = previous) }
            }
        }
    }

    /** Hides the caller's own post (web parity: own-post overflow → Hide). Same optimistic-remove +
     *  revert-on-failure as [deleteReview], but via setHidden rather than a hard delete. */
    fun hideReview(review: Review) {
        val previous = _uiState.value.reviews
        _uiState.update { it.copy(reviews = it.reviews.filterNot { r -> r.id == review.id }) }
        viewModelScope.launch {
            val result = repository.setHidden(review.id, true)
            if (result is NetworkResult.Error) {
                logger.e(TAG, "Hide review failed: ${result.error}")
                _uiState.update { it.copy(reviews = previous) }
            }
        }
    }

    /**
     * Follows / unfollows the AUTHOR of [review] without leaving the clip — the rail's "+" badge.
     * The same `POST /api/users/{id}/follow` toggle the profile screen uses (server flips the
     * row and returns `following`). Optimistic: every loaded row by that author flips at once,
     * the server's answer is written back, a failure reverts. A no-op for the viewer's own posts
     * (the server refuses `follow_self` anyway) and while a toggle for that author is in flight.
     */
    fun toggleFollow(review: Review) {
        val authorId = review.userId
        if (authorId == _uiState.value.currentUserId || authorId in followInFlight) return
        val target = !review.isFollowingAuthor
        followInFlight += authorId
        _uiState.update { s -> s.copy(reviews = s.reviews.withFollowState(authorId, target)) }
        viewModelScope.launch {
            when (val result = repository.toggleFollow(authorId)) {
                is NetworkResult.Success ->
                    _uiState.update { s -> s.copy(reviews = s.reviews.withFollowState(authorId, result.data)) }
                is NetworkResult.Error -> {
                    logger.e(TAG, "Follow toggle failed: ${result.error}")
                    _uiState.update { s -> s.copy(reviews = s.reviews.withFollowState(authorId, !target)) }
                }
            }
            followInFlight -= authorId
        }
    }

    private val followInFlight = mutableSetOf<String>()

    /**
     * The comment sheet's result: the server's comment count for ONE review, written into the feed
     * row so the rail shows it after the sheet closes (web: `addComment` in reviews/page.tsx). Only
     * that row changes; nothing is reloaded.
     */
    fun setCommentCount(reviewId: String, count: Int) {
        _uiState.update { state -> state.copy(reviews = state.reviews.withCommentCount(reviewId, count)) }
    }

    private inline fun updateReview(id: String, crossinline transform: (Review) -> Review) {
        _uiState.update { state ->
            state.copy(reviews = state.reviews.map { if (it.id == id) transform(it) else it })
        }
    }

    /**
     * One page of rows for [source]. Explore: the discovery feed with the tab's sort/following.
     * Profile: the grid's own call — `getMine()` for the signed-in user (unpaged, hidden/held posts
     * included, exactly what the Self Profile grid shows) or `getFeed(userId=…, latest)` for an
     * author (what the Author Profile grid shows). A profile source never touches the discovery
     * feed, so the pager cannot pick up another author's clip.
     */
    private suspend fun loadPage(pageIndex: Int, type: ReviewFeedType): NetworkResult<List<Review>> = when (val s = source) {
        ReviewsFeedSource.Explore ->
            repository.getFeed(page = pageIndex, limit = PAGE_SIZE, sort = sortFor(type), following = followingFor(type))
        is ReviewsFeedSource.Profile -> when (val userId = s.userId) {
            null -> if (pageIndex == 0) repository.getMine() else NetworkResult.Success(emptyList())
            else -> repository.getFeed(page = pageIndex, limit = PAGE_SIZE, sort = "latest", userId = userId)
        }
        ReviewsFeedSource.Saved -> if (pageIndex == 0) loadSaved() else NetworkResult.Success(emptyList())
    }

    /** The viewer's own unpaged lists — own posts (`/mine`) and saved (`/saved`): one call, no page 2. */
    private val ReviewsFeedSource.isMine: Boolean
        get() = (this is ReviewsFeedSource.Profile && userId == null) || this is ReviewsFeedSource.Saved

    /**
     * The saved list, playable: `GET /api/reviews/saved` answers reduced rows (tile fields only —
     * no author, media or counts), so each is fetched in full through the existing
     * `GET /api/reviews/{id}`, a few at a time, keeping the saved order. A row the server no
     * longer serves (hidden or deleted since it was saved → 404) is dropped rather than shown
     * empty; only when every row fails does the pager report the failure.
     */
    private suspend fun loadSaved(): NetworkResult<List<Review>> {
        val saved = repository.getSaved()
        if (saved !is NetworkResult.Success || saved.data.isEmpty()) return saved
        val limit = Semaphore(HYDRATE_CONCURRENCY)
        val full = coroutineScope {
            saved.data.map { row -> async { limit.withPermit { repository.getReview(row.id) } } }.awaitAll()
        }
        val rows = full.mapNotNull { (it as? NetworkResult.Success)?.data }
        if (rows.isEmpty()) full.firstOrNull { it is NetworkResult.Error }?.let { return it as NetworkResult.Error }
        return NetworkResult.Success(rows)
    }

    // Exact sort/following params the web sends per tab: For You → trending ranking; Following →
    // followed-authors, latest order; Latest → plain reverse-chronological.
    private fun sortFor(type: ReviewFeedType): String = when (type) {
        ReviewFeedType.ForYou -> "trending"
        ReviewFeedType.Latest -> "latest"
        ReviewFeedType.Following -> "latest"
    }

    private fun followingFor(type: ReviewFeedType): Boolean = type == ReviewFeedType.Following

    // ── Video watch analytics (feed-only, mirrors the web's behaviorTracker) ──────────
    private var activeVideoReviewId: String? = null
    private var activeStartMs: Long = 0L
    private val videoDurations = mutableMapOf<String, Float>()

    /** Records a clip's duration (seconds), reported by the player at READY — used to compute the
     *  completion_rate when the watch is finalized. */
    fun onVideoDuration(reviewId: String, durationSec: Float) {
        if (durationSec > 0f) videoDurations[reviewId] = durationSec
    }

    /**
     * Called when the settled feed page changes (or the first page loads). Finalizes the previously
     * active clip's watch, then starts timing [review] if it's a playable upload/link video. The
     * pager's settled page is the Android equivalent of the web tracker's ≥50%-visible clip.
     */
    fun onActiveReviewChanged(review: Review?) {
        finalizeActiveWatch()
        if (review != null && review.contentType == ReviewContentType.Video && review.mediaUrl != null) {
            activeVideoReviewId = review.id
            activeStartMs = System.currentTimeMillis()
        }
    }

    /** Finalizes any in-progress watch (e.g. when the feed leaves composition). */
    fun flushWatch() = finalizeActiveWatch()

    private fun finalizeActiveWatch() {
        val reviewId = activeVideoReviewId ?: return
        activeVideoReviewId = null
        val watchedSec = (System.currentTimeMillis() - activeStartMs) / 1000.0
        if (watchedSec < MIN_WATCH_SECONDS) return
        val duration = videoDurations[reviewId]
        val completion = if (duration != null && duration > 0f) minOf(watchedSec / duration, 1.0) else 0.0
        val watchSeconds = watchedSec.roundToInt()
        val completionRate = round(completion * 100) / 100
        viewModelScope.launch {
            // Fire-and-forget — a failure never affects playback (matches the web's sendBeacon).
            repository.recordInteraction(reviewId, watchSeconds, completionRate)
        }
    }

    private companion object {
        const val TAG = "ReviewsFeedViewModel"
        const val PAGE_SIZE = 12
        /** Saved rows hydrated in parallel, at most this many requests in flight. */
        const val HYDRATE_CONCURRENCY = 4
        const val PREFETCH_DISTANCE = 3
        const val MIN_WATCH_SECONDS = 3.0
    }
}

/**
 * The pager's source from the destination's arguments: `ReviewsRoute.ProfileClips` carries
 * `startReviewId` (+ nullable `userId`, + `saved`); the Feed destination carries nothing → Explore.
 */
internal fun sourceFrom(savedStateHandle: SavedStateHandle): ReviewsFeedSource =
    if (savedStateHandle.contains("startReviewId")) {
        val route = savedStateHandle.toRoute<ReviewsRoute.ProfileClips>()
        if (route.saved) ReviewsFeedSource.Saved else ReviewsFeedSource.Profile(userId = route.userId)
    } else {
        ReviewsFeedSource.Explore
    }

/** The pager page a profile opens on: the tapped clip's row, or the first row when it is gone. */
internal fun initialPageFor(reviews: List<Review>, startReviewId: String): Int =
    reviews.indexOfFirst { it.id == startReviewId }.coerceAtLeast(0)

/** [toggleFollow]'s row transform: every row by [authorId] gets [following]; other rows are the same object. */
internal fun List<Review>.withFollowState(authorId: String, following: Boolean): List<Review> =
    map { if (it.userId == authorId && it.isFollowingAuthor != following) it.copy(isFollowingAuthor = following) else it }

/** [setCommentCount]'s row transform: the row with [reviewId] gets [count]; every other row is the same object. */
internal fun List<Review>.withCommentCount(reviewId: String, count: Int): List<Review> =
    map { if (it.id == reviewId && it.commentCount != count) it.copy(commentCount = count) else it }

File: ios/TappyAI/Features/Reviews/UI/ReviewsFeedViewModel.swift
import Foundation
import Combine

@MainActor
final class ReviewsFeedViewModel: AppObservableObject {
    @AppPublished var reviews: [Review] = []
    @AppPublished var activeTab: FeedTab = .forYou
    @AppPublished var isLoading = false
    @AppPublished var isLoadingMore = false
    @AppPublished var error: AppError?
    @AppPublished var activeIndex: Int = 0
    @AppPublished var commentReviewId: String?
    @AppPublished var comments: [ReviewComment] = []
    @AppPublished var commentCount: Int = 0
    @AppPublished var isLoadingComments = false
    @AppPublished var commentText = ""
    @AppPublished var isPostingComment = false
    @AppPublished var commentError: String?
    @AppPublished var shareReviewId: String?
    @AppPublished var menuReviewId: String?

    let service: ReviewsService
    private let session: SessionStore
    private var page = 0
    private var hasMore = true
    private var feedTask: Task<Void, Never>?
    private let log = AppLogger.reviews

    var isAuthenticated: Bool { session.state.isAuthenticated }
    var currentUserId: String? { session.userId }

    init(service: ReviewsService, session: SessionStore) {
        self.service = service
        self.session = session
    }

    // MARK: - Feed loading

    func loadFeed() async {
        guard !isLoading else { return }
        isLoading = true
        error = nil
        page = 0
        hasMore = true

        do {
            let response = try await service.fetchFeed(
                page: 0,
                sort: feedSort,
                following: activeTab == .following
            )
            reviews = response.reviews
            hasMore = response.reviews.count >= response.limit
            page = 1
            activeIndex = 0
        } catch {
            self.error = error as? AppError ?? .unexpected(message: error.localizedDescription)
            log.error("feed load failed: \(error)")
        }
        isLoading = false
    }

    func loadMore() async {
        guard !isLoadingMore, hasMore, !isLoading else { return }
        isLoadingMore = true

        do {
            let response = try await service.fetchFeed(
                page: page,
                sort: feedSort,
                following: activeTab == .following
            )
            reviews.append(contentsOf: response.reviews)
            hasMore = response.reviews.count >= response.limit
            page += 1
        } catch {
            log.error("feed load more failed: \(error)")
        }
        isLoadingMore = false
    }

    func refresh() async {
        page = 0
        hasMore = true
        error = nil

        do {
            let response = try await service.fetchFeed(
                page: 0,
                sort: feedSort,
                following: activeTab == .following
            )
            reviews = response.reviews
            hasMore = response.reviews.count >= response.limit
            page = 1
            activeIndex = 0
        } catch {
            log.error("feed refresh failed: \(error)")
        }
    }

    func switchTab(_ tab: FeedTab) {
        guard tab != activeTab else { return }
        activeTab = tab
        reviews = []
        activeIndex = 0
        isLoading = false
        feedTask?.cancel()
        feedTask = Task { await loadFeed() }
    }

    private var feedSort: FeedSort {
        switch activeTab {
        case .forYou: return .trending
        case .following, .latest: return .latest
        }
    }

    // MARK: - Like (optimistic, matches Web double-tap + rail behavior)

    func toggleLike(reviewId: String) {
        guard let idx = reviews.firstIndex(where: { $0.id == reviewId }) else { return }
        let wasLiked = reviews[idx].likedByMe
        let oldCount = reviews[idx].likeCount
        reviews[idx].likedByMe = !wasLiked
        reviews[idx].likeCount = max(0, oldCount + (wasLiked ? -1 : 1))

        Task {
            do {
                let result = try await service.toggleLike(reviewId: reviewId)
                if let i = reviews.firstIndex(where: { $0.id == reviewId }) {
                    reviews[i].likedByMe = result.liked
                }
            } catch {
                if let i = reviews.firstIndex(where: { $0.id == reviewId }) {
                    reviews[i].likedByMe = wasLiked
                    reviews[i].likeCount = oldCount
                }
                log.error("like failed: \(error)")
            }
        }
    }

    func doubleTapLike(reviewId: String) {
        guard let idx = reviews.firstIndex(where: { $0.id == reviewId }) else { return }
        if reviews[idx].likedByMe { return }
        let oldCount = reviews[idx].likeCount
        reviews[idx].likedByMe = true
        reviews[idx].likeCount = oldCount + 1

        Task {
            do {
                let result = try await service.toggleLike(reviewId: reviewId)
                if let i = reviews.firstIndex(where: { $0.id == reviewId }) {
                    reviews[i].likedByMe = result.liked
                }
            } catch {
                if let i = reviews.firstIndex(where: { $0.id == reviewId }) {
                    reviews[i].likedByMe = false
                    reviews[i].likeCount = oldCount
                }
            }
        }
    }

    // MARK: - Save (optimistic)

    func toggleSave(reviewId: String) {
        guard let idx = reviews.firstIndex(where: { $0.id == reviewId }) else { return }
        let wasSaved = reviews[idx].savedByMe
        let oldCount = reviews[idx].saveCount
        reviews[idx].savedByMe = !wasSaved
        reviews[idx].saveCount = max(0, oldCount + (wasSaved ? -1 : 1))

        Task {
            do {
                let result = try await service.toggleSave(reviewId: reviewId)
                if let i = reviews.firstIndex(where: { $0.id == reviewId }) {
                    reviews[i].savedByMe = result.saved
                }
            } catch {
                if let i = reviews.firstIndex(where: { $0.id == reviewId }) {
                    reviews[i].savedByMe = wasSaved
                    reviews[i].saveCount = oldCount
                }
                log.error("save failed: \(error)")
            }
        }
    }

    // MARK: - Comments

    func openComments(reviewId: String) {
        commentReviewId = reviewId
        comments = []
        commentCount = 0
        commentText = ""
        commentError = nil
        Task { await loadComments(reviewId: reviewId) }
    }

    func closeComments() {
        commentReviewId = nil
    }

    func loadComments(reviewId: String) async {
        isLoadingComments = true
        commentError = nil
        do {
            let response = try await service.fetchComments(reviewId: reviewId)
            comments = response.comments
            commentCount = response.count
        } catch {
            commentError = "Không thể tải bình luận"
            log.error("comments load failed: \(error)")
        }
        isLoadingComments = false
    }

    func postComment() {
        guard let reviewId = commentReviewId else { return }
        let body = commentText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !body.isEmpty, body.count <= 300 else { return }
        isPostingComment = true
        commentError = nil
        let savedText = commentText
        commentText = ""

        Task {
            do {
                let response = try await service.postComment(reviewId: reviewId, body: body)
                comments.append(response.comment)
                commentCount = response.count
                if let idx = reviews.firstIndex(where: { $0.id == reviewId }) {
                    reviews[idx].commentCount = response.count
                }
            } catch {
                commentText = savedText
                commentError = "Không thể đăng bình luận"
                log.error("post comment failed: \(error)")
            }
            isPostingComment = false
        }
    }

    func deleteComment(commentId: String) {
        guard let reviewId = commentReviewId else { return }
        Task {
            do {
                let response = try await service.deleteComment(reviewId: reviewId, commentId: commentId)
                comments.removeAll { $0.id == commentId }
                commentCount = response.count
                if let idx = reviews.firstIndex(where: { $0.id == reviewId }) {
                    reviews[idx].commentCount = response.count
                }
            } catch {
                log.error("delete comment failed: \(error)")
            }
        }
    }

    // MARK: - Follow

    func toggleFollow(userId: String) {
        Task {
            do {
                _ = try await service.toggleFollow(userId: userId)
            } catch {
                log.error("follow failed: \(error)")
            }
        }
    }

    // MARK: - Share

    func openShare(reviewId: String) {
        shareReviewId = reviewId
    }

    func closeShare() {
        shareReviewId = nil
    }

    // MARK: - Delete / Hide (own-post menu)

    func deleteReview(reviewId: String) {
        Task {
            do {
                try await service.deleteReview(reviewId: reviewId)
                reviews.removeAll { $0.id == reviewId }
                if activeIndex >= reviews.count { activeIndex = max(0, reviews.count - 1) }
            } catch {
                log.error("delete review failed: \(error)")
            }
        }
    }

    func hideReview(reviewId: String) {
        Task {
            do {
                try await service.hideReview(reviewId: reviewId, hidden: true)
                reviews.removeAll { $0.id == reviewId }
                if activeIndex >= reviews.count { activeIndex = max(0, reviews.count - 1) }
            } catch {
                log.error("hide review failed: \(error)")
            }
        }
    }

    // MARK: - Interact (watch-time signal)

    func interact(reviewId: String, watchSeconds: Int, completionRate: Double) async {
        await service.interact(reviewId: reviewId, watchSeconds: watchSeconds, completionRate: completionRate)
    }

    // MARK: - Infinite scroll check

    func checkLoadMore(currentIndex: Int) {
        let threshold = reviews.count - 3
        if currentIndex >= threshold {
            Task { await loadMore() }
        }
    }
}
